AkiraMD/ruby-senior-take-home-assignment
# Db/interview/ruby take home

**Summary** 

This PR expands upon the existing Patients REST API Service by implementing new features and enhancements as per the THVC take-home assignment requirements. The changes primarily focus on enhancing the /patients endpoint and integrating with external vendor APIs to fetch patient records. Additionally, a caching strategy has been implemented to improve performance. Since it's a 2 to 3 hours test, I limited myself and added less tests than I would normally do.

**Changes Made**

1. Enhanced /patients/:patient_id/record Endpoint:
- Added a new endpoint to retrieve detailed patient records including province, allergies, and number of medical visits from external vendor APIs. Implemented handling for two different vendor integrations based on records_vendor and vendor_id.
2. Integration with Vendor APIs:
- Implemented logic to fetch patient records from Vendor One and Vendor Two APIs, ensuring authentication using mock tokens and adhering to API specifications provided in the assignment.
3. Caching Strategy Implemented:
- Introduced a caching mechanism using Redis and a custom caching utility (Vandelay::Util::Cache) to store fetched patient records for improved performance. Records are cached for up to 10 minutes to reduce network overhead.
4. Test Framework Integration:
- Added comprehensive RSpec tests to validate functionality across different scenarios: Testing cache retrieval and storage.

**Next Steps**
The implementation meets the basic requirements outlined in the assignment. Future enhancements could focus on:
- Adding many more tests.
- Enhancing error handling and logging mechanisms.
- Integrating real authentication and error handling for live API environments.

**Testing Instructions**
To test the changes:
- Clone the repository and switch to the branch db/interview/ruby-take-home.
- Ensure Docker is installed and running.
- Run `docker compose up -d --build` to build and start the Docker containers.
- Access the API at `http://localhost:3087` to test the implemented endpoints with the correct token (suggest to use Postman).
- Run `bundle exec rspec` to execute the test suite and verify functionality.
